packages/@aws-cdk/aws-servicecatalog/lib/portfolio.ts

```diff
@@ -0,0 +1,246 @@
+import * as iam from '@aws-cdk/aws-iam';
+import { IResource, Names, Resource, Stack } from '@aws-cdk/core';
+import { AcceptLanguage } from './common';
+import { getIdentifier } from './private/util';
+import { InputValidator } from './private/validation';
+import { CfnPortfolio, CfnPortfolioPrincipalAssociation, CfnPortfolioShare } from './servicecatalog.generated';
+
+// keep this import separate from other imports to reduce chance for merge conflicts with v2-main
+// eslint-disable-next-line no-duplicate-imports, import/order
+import { Construct } from 'constructs';
+
+/**
+ * A Service Catalog portfolio.
+ */
+export interface IPortfolio extends IResource {
+
+  /**
+   * The ARN of the portfolio.
+   * @attribute
+   */
+  readonly portfolioArn: string;
+
+  /**
+   * The ID of the portfolio.
+   * @attribute
+   */
+  readonly portfolioId: string;
+
+  /**
+   * Associate portfolio with an IAM Role.
+   * @param role an IAM role
+   */
+  giveAccessToRole(role: iam.IRole): void;
+
+  /**
+   * Associate portfolio with an IAM User.
+   * @param user an IAM user
+   */
+  giveAccessToUser(user: iam.IUser): void;
+
+  /**
+   * Associate portfolio with an IAM Group.
+   * @param group an IAM Group
+   */
+  giveAccessToGroup(group: iam.IGroup): void;
+
+  /**
+   * Share portfolio with another account.
+   * @param accountId AWS account to share portfolio with
+   * @param options Options for the initiate share
+   */
+  share(accountId: string, options?: PortfolioShareOptions): void;
+}
+
+/**
+ * Represents a Service Catalog Portfolio.
+ */
+abstract class PortfolioBase extends Resource implements IPortfolio {
+
+  /**
+   * The ARN of the portfolio.
+   */
+  public abstract readonly portfolioArn: string;
+
+  /**
+   * The Id of the portfolio.
+   */
+  public abstract readonly portfolioId: string;
+
+  /**
+   * Enable the Portfolio for a specific Role
+   */
+  public giveAccessToRole(role: iam.IRole) {
+    this.associatePrincipal(role.roleArn, role.node.addr);
+  }
+
+  /**
+   * Enable the Portfolio for a specific User
+   */
+  public giveAccessToUser(user: iam.IUser) {
+    this.associatePrincipal(user.userArn, user.node.addr);
+  }
+
+  /**
+   * Enable the Portfolio for a specific Group
+   */
+  public giveAccessToGroup(group: iam.IGroup) {
```

Comment:
All of these `giveAccessTo*()` methods should be idempotent, right? So calling

```ts
portfolio.giveAccessToGroup(group);
portfolio.giveAccessToGroup(group);
```

Should succeed, and that second call does nothing.

I believe this is not the case today, correct? Can you change that, and write some unit tests that confirm this works?

Reply:
yes, will do

Reply:
added unit test.